File: ios/CellularChatCore/Sources/CellularChatCore/Derivations.swift
import Foundation

/// Pair-specific key derivations (PROTOCOL_V2.md §2). All outputs 32 bytes
/// except the fingerprint display string.
public enum Derivations {

    public static func pairingPsk(secret: [UInt8]) -> [UInt8] {
        Crypto.hmacSHA256(key: secret, data: Array("cellfind/v2 pairing psk".utf8))
    }

    /// pairRoot = HKDF(salt = h_pairing, ikm = pairingPsk,
    ///                 info = "cellfind/v2 pair root" || staticA || staticB)
    public static func pairRoot(pairingHandshakeHash: [UInt8],
                                pairingPsk: [UInt8],
                                staticA: [UInt8],
                                staticB: [UInt8]) -> [UInt8] {
        let info = Array("cellfind/v2 pair root".utf8) + staticA + staticB
        return Crypto.hkdfRFC5869(salt: pairingHandshakeHash, ikm: pairingPsk, info: info)
    }

    public static func sessionPsk(pairRoot: [UInt8]) -> [UInt8] {
        Crypto.hmacSHA256(key: pairRoot, data: Array("cellfind/v2 session psk".utf8))
    }

    public static func discoveryKey(pairRoot: [UInt8], role: PairRole) -> [UInt8] {
        let label = role == .a ? "cellfind/v2 discovery A" : "cellfind/v2 discovery B"
        return Crypto.hmacSHA256(key: pairRoot, data: Array(label.utf8))
    }

    public static func confirmMAC(pairRoot: [UInt8], role: PairRole) -> [UInt8] {
        let roleByte: UInt8 = role == .a ? 0x41 : 0x42
        return Crypto.hmacSHA256(
            key: pairRoot, data: Array("cellfind/v2 confirm".utf8) + [roleByte])
    }

    public static func fingerprint(pairId: [UInt8], staticA: [UInt8], staticB: [UInt8]) -> [UInt8] {
        Crypto.sha256(Array("cellfind/v2 fingerprint".utf8) + pairId + staticA + staticB)
    }

    /// decimal(u64BE(fpr[0..8]) mod 1_000_000), zero-padded to 6 digits.
    public static func fingerprintDisplay(pairId: [UInt8], staticA: [UInt8], staticB: [UInt8]) -> String {
        let fpr = fingerprint(pairId: pairId, staticA: staticA, staticB: staticB)
        var value: UInt64 = 0
        for i in 0..<8 { value = (value << 8) | UInt64(fpr[i]) }
        let digits = value % 1_000_000
        return String(format: "%06d", digits)
    }
}

/// Role A = inviter, Role B = joiner (PROTOCOL_V2.md §2). Permanent per pair.
public enum PairRole: Equatable {
    case a
    case b
}

File: android/app/src/main/java/com/cellularchat/app/core/crypto/Derivations.kt
package com.cellularchat.app.core.crypto

import java.nio.charset.StandardCharsets

/** Pair-specific key derivations (PROTOCOL_V2.md §2). All outputs are 32 bytes. */
object Derivations {
    const val ROLE_A: Byte = 0x41
    const val ROLE_B: Byte = 0x42

    fun pairingPsk(secret: ByteArray): ByteArray =
        Hashes.hmacSha256(secret, ascii("cellfind/v2 pairing psk"))

    fun pairRoot(pairingHandshakeHash: ByteArray, pairingPsk: ByteArray, staticA: ByteArray, staticB: ByteArray): ByteArray =
        Hashes.hkdf(
            salt = pairingHandshakeHash,
            ikm = pairingPsk,
            info = ascii("cellfind/v2 pair root") + staticA + staticB,
        )

    fun sessionPsk(pairRoot: ByteArray): ByteArray =
        Hashes.hmacSha256(pairRoot, ascii("cellfind/v2 session psk"))

    fun discoveryKey(pairRoot: ByteArray, role: Byte): ByteArray {
        val label = if (role == ROLE_A) "cellfind/v2 discovery A" else "cellfind/v2 discovery B"
        return Hashes.hmacSha256(pairRoot, ascii(label))
    }

    fun confirmMac(pairRoot: ByteArray, role: Byte): ByteArray =
        Hashes.hmacSha256(pairRoot, ascii("cellfind/v2 confirm") + role)

    fun fingerprint(pairId: ByteArray, staticA: ByteArray, staticB: ByteArray): ByteArray =
        Hashes.sha256(ascii("cellfind/v2 fingerprint") + pairId + staticA + staticB)

    /** Six-digit display code from the fingerprint (§2). */
    fun fingerprintDisplay(fingerprint: ByteArray): String {
        var value = 0L
        for (i in 0 until 8) value = (value shl 8) or (fingerprint[i].toLong() and 0xff)
        val mod = java.lang.Long.remainderUnsigned(value, 1_000_000L)
        return mod.toString().padStart(6, '0')
    }

    private fun ascii(text: String): ByteArray = text.toByteArray(StandardCharsets.US_ASCII)
}
